src/bytes.js

```diff
@@ -0,0 +1,1497 @@
+        }
+        if (!Sk.builtin.checkString(pySource)) {
+            // think ahead for kwarg support
+            throw new Sk.builtin.TypeError((encoding !== undefined ? "encoding" : "errors") + " without a string argument");
+        }
+    }
+
+    if (pySource === undefined) {
+        return new Sk.builtin.bytes();
+    } else if (Sk.builtin.checkString(pySource)) {
+        if (encoding === undefined) {
+            throw new Sk.builtin.TypeError("string argument without an encoding");
+        }
+        errors = errors === undefined ? "strict" : errors.$jsstr();
+        encoding = encoding.$jsstr();
+        return strEncode(pySource, encoding, errors);
+    } else if (Sk.builtin.checkInt(pySource)) {
+        source = Sk.builtin.asnum$(pySource);
+        if (source < 0) {
+            throw new Sk.builtin.ValueError("negative count");
+        } else if (source > Number.MAX_SAFE_INTEGER) {
+            throw new Sk.builtin.OverflowError("cannot fit 'int' into an index-sized integer");
+        } 
+        return new Sk.builtin.bytes(source);
+    } else if (Sk.builtin.checkBytes(pySource)) {
+        return new Sk.builtin.bytes(pySource.v);
+    } else if ((dunderBytes = Sk.abstr.lookupSpecial(pySource, Sk.builtin.str.$bytes)) !== undefined) {
+        const ret = Sk.misceval.callsimOrSuspendArray(dunderBytes, [pySource]);
+        return Sk.misceval.chain(ret, (bytesSource) => {
+            if (!Sk.builtin.checkBytes(bytesSource)) {
+                throw new Sk.builtin.TypeError("__bytes__ returned non-bytes (type " + Sk.abstr.typeName(bytesSource) + ")");
+            }
+            return bytesSource;
+        });
+    } else if (Sk.builtin.checkIterable(pySource)) {
+        source = [];
+        const r = Sk.misceval.iterFor(Sk.abstr.iter(pySource), (byte) => {
+            if (!Sk.misceval.isIndex(byte)) {
+                throw new Sk.builtin.TypeError("'" + Sk.abstr.typeName(byte) + "' object cannot be interpreted as an integer");
+            };
+            const n = Sk.misceval.asIndex(byte);
+            if (n < 0 || n > 255) {
+                throw new Sk.builtin.ValueError("bytes must be in range(0, 256)");
+            }
+            source.push(n);
+        });
+        return Sk.misceval.chain(r, () => new Sk.builtin.bytes(source));
+    }
+    throw new Sk.builtin.TypeError("cannot convert '" + Sk.abstr.typeName(pySource) + "' object into bytes");
+}
+
+function makehexform(num) {
+    var leading;
+    if (num <= 265) {
+        leading = "\\x";
+    } else {
+        leading = "\\u";
+    }
+    num = num.toString(16);
+    if (num.length === 3) {
+        num = num.slice(1, 3);
+    }
+    if (num.length === 1) {
+        num = leading + "0" + num;
+    } else {
+        num = leading + num;
+    }
+    return num;
+};
+
+
+Sk.builtin.bytes.prototype.$jsstr = function () {
+    return Decoder.decode(this.v);
```

Comment:
This will throw an error if the bytes object isn't valid UTF-8. In the contexts in which this is used, I suggest it would be more helpful to return a JS binary string (a JS string with character codes in the range 0-255). This will also make life rather easier for those of us who have to accept bytestrings from Python 2 and 3!

Reply:
adjusted to loop throw and return the `String.fromCharCode`